File: pages/api/getData.js
// import puppeteer from "puppeteer";
import { google } from "googleapis";
import createClient from "../../lib/createClient";

let chrome = {};
let puppeteer;

if (process.env.VERCEL) {
  // running on the Vercel platform.
  chrome = require("chrome-aws-lambda");
  puppeteer = require("puppeteer-core");
} else {
  // running locally.
  puppeteer = require("puppeteer");
}

const calendar = google.calendar("v3");
const HOW_MANY_WEEKS = 10;

export default async function GetData(req, res) {
  const USER_ID = req.body.ID;
  const USER_PW = req.body.PW;
  const CODE = req.body.CODE;
  try {
    //클라이언트 생성, 토큰 획득
    const oAuth2Client = createClient();
    oAuth2Client.getToken(CODE, (err, token) => {
      if (err) throw "Error retrieving access token: " + err;
      oAuth2Client.setCredentials(token);
    });

    google.options({ auth: oAuth2Client });
    let schedules = await NockingWebsite(USER_ID, USER_PW);
    // result = await calendar.calendarList.list();
    await insertEvent(schedules);
    res.redirect("/complete");
  } catch (err) {
    console.error(err);
    res.writeHead(200, { "Content-Type": "text/plain;charset=utf-8" });
    res.write(err);
    res.end();
  }
}

const NockingWebsite = async (id, pw) => {
  // const browser = await puppeteer.launch({ headless: true });
  const browser = await puppeteer.launch({
    args: [...chrome.args, "--hide-scrollbars", "--disable-web-security"],
    defaultViewport: chrome.defaultViewport,
    executablePath: await chrome.executablePath,
    headless: true,
    ignoreHTTPSErrors: true,
  });

  const page = await browser.newPage();
  await page.goto("https://case.publicdef.net");

  const loginElem = await page.$("div.navbar > a");
  await loginElem.click();

  const userNameElem = await page.$("input#id_username");
  await userNameElem.type(id);

  const passwordElem = await page.$("input#id_password");
  await passwordElem.type(pw);

  const loginButton = await page.$("input#loginbutton");
  await Promise.all([page.waitForNavigation(), loginButton.click()]);
  if (page.url().includes("login")) {
    throw "로그인에 실패하였습니다.";
  }

  const monthlySchedule = await page.$("div.navbar > a");
  await Promise.all([page.waitForNavigation(), monthlySchedule.click()]);

  const weeklySchedule = await page.$(".fc-listWeek-button");
  await weeklySchedule.click();

  let schedules = await getSchedules(page);
  for (let i = 0; i < HOW_MANY_WEEKS; i++) {
    const next = await page.$(".fc-next-button");
    await next.click();
    let temp = await getSchedules(page);
    schedules = [...schedules, ...temp];
  }
  return schedules;
};

const getSchedules = async (page) => {
  let result = await page.$$eval(".fc-list-table > tbody > tr", (trs) => {
    let date;
    let arr = [];
    for (tr of trs) {
      if (tr.className === "fc-list-heading") {
        date = tr.dataset.date;
        continue;
      }
      const content = tr.querySelector("td.fc-list-item-title").textContent;
      if (content.includes("기일표")) {
        continue;
      }
      const time = tr.querySelector("td.fc-list-item-time").textContent;
      arr.push({ date, time, content });
    }
    return arr;
  });
  return result;
};

async function insertEvent(schedules) {
  try {
    let oldCalendars = (await calendar.calendarList.list()).data.items;
    if (oldCalendars !== null) {
      let oldGookseons = oldCalendars.filter((item) => item.summary === "국선");
      for (let e of oldGookseons) {
        calendar.calendars.delete({ calendarId: e.id });
      }
    }
    const newCalendar = await calendar.calendars.insert({
      requestBody: {
        summary: "국선",
        timeZone: "Asia/Seoul",
      },
    });

    // let tasks = schedules.map(async (schedule) => {
    //   let event = { start: {}, end: {}, summary: "" };
    //   event.summary = schedule.content;
    //   if (schedule.time == "종일") {
    //     event.start.date = schedule.date;
    //     event.end.date = schedule.date;
    //   } else {
    //     event.start.dateTime = new Date(schedule.date + " " + schedule.time);
    //     event.end.dateTime = new Date(schedule.date + " " + schedule.time);
    //   }
    //   await sleep(500);
    //   return calendar.events.insert({
    //     calendarId: newCalendar.data.id,
    //     requestBody: event,
    //   });
    // });
    // await Promise.all(tasks);
    // console.log("완료!!!");

    for (let i = 0; i < schedules.length; i++) {
      let event = { start: {}, end: {}, summary: "" };

      event.summary = schedules[i]["content"];
      if (schedules[i].time == "종일") {
        event.start.date = schedules[i].date;
        event.end.date = schedules[i].date;
      } else {
        event.start.dateTime = new Date(
          schedules[i].date + " " + schedules[i].time
        );
        event.end.dateTime = new Date(
          schedules[i].date + " " + schedules[i].time
        );
      }
      console.log(event);
      let createEvent = await calendar.events.insert({
        calendarId: newCalendar.data.id,
        requestBody: event,
      });
      if (createEvent.statusText == "OK") {
        console.log("사건 일정이 구글캘린더에 추가되었습니다.");
      } else {
        console.log("사건 일정을 구글캘린더에 추가하는 데 실패하였습니다.");
      }
    }
  } catch (e) {
    console.error(e);
  }
}

File: lib/functions/get_data.dart
import 'package:get/get.dart';
import 'package:puppeteer/puppeteer.dart' as ppt;
import '../controller/message_controller.dart';

var mc = Get.find<MessageController>();

const baseURL = 'https://case.publicdef.net';

Future<List<dynamic>?> getData({
  required String id,
  required String password,
  required int howManyWeeks,
}) async {
  var browser = await ppt.puppeteer.launch(headless: true);
  var myPage = await browser.newPage();
  await myPage.goto(baseURL);
  var loginElem = await myPage.$('div.navbar > a');
  await loginElem.click();
  var userNameElem = await myPage.$('input#id_username');
  await userNameElem.type(id);
  var passwordElem = await myPage.$('input#id_password');
  await passwordElem.type(password);
  var loginButton = await myPage.$('input#loginbutton');
  await Future.wait(
    [myPage.waitForNavigation(), loginButton.click()],
  );

  if (myPage.url!.contains('login')) {
    mc.changeText(mc.messages.notLoaded);
    return null;
  }
  mc.changeText(mc.messages.logined);

  var monthlySchedule = await myPage.$('div.navbar > a');
  await Future.wait(
    [myPage.waitForNavigation(), monthlySchedule.click()],
  );

  var weeklySchedule = await myPage.$('.fc-listWeek-button');
  await weeklySchedule.click();

  mc.changeText(mc.messages.loading);

  //주별 일정 불러오기
  List<dynamic> schedules = await getSchedules(myPage);
  for (int i = 0; i < howManyWeeks; i++) {
    var next = await myPage.$('.fc-next-button');
    await next.click();
    var temp = await getSchedules(myPage);
    schedules = [...schedules, ...temp];
  }

  //법정 추가
  for (var i = 0; i < schedules.length; i++) {
    if (!schedules[i]['link'].contains('사건상세내용')) {
      continue;
    }
    await Future.wait([
      myPage.waitForNavigation(),
      myPage.goto(baseURL + schedules[i]['link']),
      myPage.waitForSelector('tbody.casedetail_td')
    ]);
    String courtroom = await myPage.$eval('tbody.casedetail_td', r'''
        node => {
          var arr = node.innerText.match(/\d{3,}호/g); 
          return arr[arr.length - 1];
          }
      ''');
    // RegExp exp = RegExp(r'\d+호');
    // courtroom = '[' + (exp.stringMatch(courtroom) ?? '?') + ']';
    courtroom = '[' + courtroom + ']';
    schedules[i]['courtroom'] = courtroom;
  }

  mc.changeText(mc.messages.oauth);

  // for (final elem in schedules) {
  //   print(elem);
  // }

  return schedules;
}

Future<List<dynamic>> getSchedules(ppt.Page page) async {
  List<dynamic> result = await page.$$eval('.fc-list-table > tbody > tr', '''
    function (trs) { 
      var date;
      var arr = [];
      for(tr of trs) {
        if(tr.className === 'fc-list-heading') {
          date = tr.dataset.date;
          continue;
        }
        var content = tr.querySelector('td.fc-list-item-title').textContent;
        var time = tr.querySelector('td.fc-list-item-time').textContent;
        var link = tr.querySelector('a').getAttribute('href');

        if(content.includes('기일표') || content.includes('의견서')) {
          continue;
        }     
        
        arr.push({date, time, content, link});
      }
      return arr;
    }
    ''');

  return result;
}
// if(content.includes('기일표') || content.includes('의견서') || !link.includes('사건상세내용')) {
//           continue;
//         }